### PR TITLE
New `LetterBox(size)` `CenterCrop(size)`, `ToTensor()` transforms (#9213)

### LINKED ISSUE #9192: Consistent and Efficient Preprocessing for Classification and Detection Model
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

The new classification model utilizes a different preprocessing pipeline as the detection model. For object detection, the image is normalized, resized with constant aspect ratio and padded to size. In comparison, the new classification models utilizes normalization, resizing, center crop and standardization by ImageNet mean and std.

### Use case

We use YOLOv5 models for embedded applications with our own performance optimized software framework. Due to the two pipelines, the models are not easily interchangeable. Furthermore, the classification preprocessing pipeline is more performance intensive and therefore not well suited for low power environments.

Would it be possible to introduce a compatibility flag or some other solution to export classification models which expect the same normalized rgb image input?

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by glenn-jocher**
YOLOv5 LetterBox class for image preprocessing, i.e. T.Compose([T.ToTensor(), LetterBox(size)])

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

@AyushExel 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to image preprocessing in YOLOv5 with new transformation classes and modified dataloader behavior.

### 📊 Key Changes
- 📝 Introduced `LetterBox`, `CenterCrop`, and `ToTensor` custom preprocessing classes in `augmentations.py`.
- 👁️‍🗨️ Modified the behavior of default transformations to use new custom classes instead of torchvision transforms.
- 🔄 Updated `dataloaders.py` to incorporate the new transformation classes and streamlined image color conversion when transforms are applied.
- 🗂️ Cache improvements for storing and retrieving processed images, reducing I/O operations and potentially speeding up training.

### 🎯 Purpose & Impact
- 💡 **Purpose**: To provide YOLOv5 with more flexible image preprocessing options, specifically tailored to the YOLO architecture, and improve performance through better caching mechanisms.
- 🚀 **Impact**: Expect increased efficiency in data loading, a potential boost in training speed, and easier customization of image preprocessing steps for YOLOv5 users.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@AyushExel this replicates our detection pre-processing. It's maybe 10% faster than current cls preprocessing. Our existing transforms are way faster than either though:

```python

            TF = T.Compose([T.ToTensor(), T.Resize(size), T.CenterCrop(size), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])
            with Profile() as dt:
                for _ in range(1000):
                    im = TF(cv2.cvtColor(im0, cv2.COLOR_BGR2RGB))
            print(dt.t)

            TF = T.Compose([T.ToTensor(), LetterBox(size)])
            with Profile() as dt:
                for _ in range(1000):
                    im = TF(cv2.cvtColor(im0, cv2.COLOR_BGR2RGB))
            print(dt.t)

            with Profile() as dt:
                for _ in range(1000):
                    im = letterbox(im0, self.img_size, stride=self.stride, auto=self.auto)[0]  # padded resize
                    im = im.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB
                    im = np.ascontiguousarray(im)  # contiguous
                    im = torch.from_numpy(im).to('cpu')
                    im = im.half() if False else im.float()  # uint8 to fp16/32
                    im /= 255  # 0 - 255 to 0.0 - 1.0
            print(dt.t)

# 5.506530046463013
# 4.960566997528076
# 1.5980091094970703
```

**Comment by glenn-jocher** [comment]
Just experimenting right now to see what effect the different preprocessing has on imagewoof.

**Comment by glenn-jocher** [comment]
@AyushExel wow this is super awesome. If I replace the torch transforms with the ones I made our ImageNet val accuracies stay the same but our total time goes from 1:36 to 1:07, about 30% faster!

### Master
```
T.Compose([T.ToTensor(), T.Resize(size), T.CenterCrop(size), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])

validating: 100% 391/391 [01:36<00:00,  4.07it/s]
                   Class      Images    top1_acc    top5_acc
                     all       50000       0.759       0.928
```

### PR
```
T.Compose([CenterCrop(size), ToTensor(), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])

validating: 100% 391/391 [01:07<00:00,  5.83it/s]
                   Class      Images    top1_acc    top5_acc
                     all       50000       0.759       0.929
```

**Comment by AyushExel** [comment]
@glenn-jocher okay this is great! But let's not merge it before testing on imagenet. 92% accuracy is too high so maybe it's not the right dataset to confirm performance variation.

**Comment by glenn-jocher** [comment]
@AyushExel oh that comparison is already on ImageNet. It's using the official models on the 50,000 imagenet val images using
```python
!bash yolov5-glenn/data/scripts/get_imagenet.sh --val  # download ImageNet val split (6.3G - 50000 images)
!python yolov5/classify/val.py --weights yolov5m-cls.pt --data ../datasets/imagenet --img 224  # validate
```

**Comment by glenn-jocher** [comment]
Ok PR is all good. Tested on Imagewoof and Imagenette training. Same accuracy but about 10-20% faster training times, i.e. 24 min vs 29 min. 

Screenshot 2022-08-30 at 15 14 04
